internal/txapp/routes.go

```diff
@@ -156,92 +201,117 @@ func (d *deployDatasetRoute) Execute(ctx TxContext, router *TxApp, tx *transacti
 }
 
-func (d *dropDatasetRoute) Price(ctx context.Context, router *TxApp, tx *transactions.Transaction) (*big.Int, error) {
+func (d *dropDatasetRoute) Price(ctx context.Context, app *common.App, tx *transactions.Transaction) (*big.Int, error) {
 	return big.NewInt(10000000000000), nil
 }
 
-type executeActionRoute struct{}
-
-func (e *executeActionRoute) Execute(ctx TxContext, router *TxApp, tx *transactions.Transaction) *TxResponse {
-	dbTx, err := router.currentTx.BeginTx(ctx.Ctx)
+func (d *dropDatasetRoute) PreTx(ctx common.TxContext, app *common.App, tx *transactions.Transaction) (transactions.TxCode, error) {
+	drop := &transactions.DropSchema{}
+	err := drop.UnmarshalBinary(tx.Body.Payload)
 	if err != nil {
-		return txRes(nil, transactions.CodeUnknownError, err)
+		return transactions.CodeEncodingError, err
 	}
 
-	spend, code, err := router.checkAndSpend(ctx, tx, e, dbTx)
+	d.dbid = drop.DBID
+	return 0, nil
+}
+
+func (d *dropDatasetRoute) InTx(ctx common.TxContext, app *common.App, tx *transactions.Transaction) (transactions.TxCode, error) {
+	err := app.Engine.DeleteDataset(ctx.Ctx, app.DB, d.dbid, tx.Sender)
 	if err != nil {
-		switch code {
-		case transactions.CodeOk, transactions.CodeInsufficientBalance, transactions.CodeInsufficientFee:
-			logErr(router.log, dbTx.Commit(ctx.Ctx))
-		default:
-			logErr(router.log, dbTx.Rollback(ctx.Ctx))
-		}
-		return txRes(spend, code, err)
+		return transactions.CodeUnknownError, err
 	}
-	defer dbTx.Commit(ctx.Ctx)
+	return 0, nil
+}
+
+type executeActionRoute struct {
+	identifier string
+	dbid       string
+	action     string
+	args       [][]any
+}
+
+var _ consensus.Route = (*executeActionRoute)(nil)
 
+func (d *executeActionRoute) Name() string {
+	return transactions.PayloadTypeExecuteAction.String()
+}
+
+func (d *executeActionRoute) Price(ctx context.Context, app *common.App, tx *transactions.Transaction) (*big.Int, error) {
+	return big.NewInt(2000000000000000), nil
+}
+
+func (d *executeActionRoute) PreTx(ctx common.TxContext, app *common.App, tx *transactions.Transaction) (transactions.TxCode, error) {
 	action := &transactions.ActionExecution{}
-	err = action.UnmarshalBinary(tx.Body.Payload)
+	err := action.UnmarshalBinary(tx.Body.Payload)
 	if err != nil {
-		return txRes(spend, transactions.CodeEncodingError, err)
+		return transactions.CodeEncodingError, err
 	}
 
-	identifier, err := ident.Identifier(tx.Signature.Type, tx.Sender)
-	if err != nil {
-		return txRes(spend, transactions.CodeUnknownError, err)
+	if len(action.NilArg) > len(action.Arguments) {
+		return transactions.CodeEncodingError, errors.New("too many NilArgs")
 	}
 
-	if len(action.NilArg) > len(action.Arguments) {
-		return txRes(nil, transactions.CodeEncodingError, errors.New("too many NilArgs"))
+	d.action = action.Action
+	d.dbid = action.DBID
+
+	d.identifier, err = ident.Identifier(tx.Signature.Type, tx.Sender)
+	if err != nil {
+		return transactions.CodeUnknownError, err
```

Comment:
Another place to return a better code.  This is 100% a problem with either the sig type or the sender/pubkey, and code should indicate.